libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt

```diff
@@ -0,0 +1,94 @@
+package com.mapbox.navigation.dropin.component.maneuver
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import com.mapbox.navigation.base.formatter.DistanceFormatter
+import com.mapbox.navigation.ui.maneuver.api.MapboxManeuverApi
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.scan
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.launch
+
+@ExperimentalCoroutinesApi
+class ManeuverViewModel(
+    private val distanceFormatter: DistanceFormatter
+) : ViewModel() {
+
+    private val maneuverApi = MapboxManeuverApi(distanceFormatter)
+    private val _maneuverState: MutableStateFlow<ManeuverState> =
+        MutableStateFlow(ManeuverState.initial())
+
+    internal fun maneuverState(): Flow<ManeuverState> = _maneuverState
+
+    internal fun consumeAction(action: Flow<ManeuverAction>) {
+        viewModelScope.launch {
+            action
+                .toResult()
+                .reduce()
+                .onEach { _maneuverState.value = it }
+                .stateIn(viewModelScope)
+        }
+    }
+
+    private fun Flow<ManeuverAction>.toResult(): Flow<ManeuverResult> =
+        map { value: ManeuverAction ->
+            when (value) {
+                is ManeuverAction.UpdateNavigationState -> {
+                    val processor = ManeuverProcessor.ProcessNavigationState(value.navigationState)
+                    processor.process()
+                }
+                is ManeuverAction.UpdateVisibility -> {
+                    val processor = ManeuverProcessor.ProcessVisibility(
+                        _maneuverState.value.navigationState
+                    )
+                    processor.process()
```

Comment:
We don't need a distinct "update visibility" action. Visibility is based on state and data which are both separate actions.

Reply:
Same everywhere below.

Reply:
I had it exactly like that first where a change in `NavigationState` would also update the visibility. Although it makes complete sense, I think we loose some control here and out of nowhere without any intention changing the state changes the visibility as well. I agree it is probably verbose, but it makes more sense to me trigger the action to update visibility rather than some other action that updates a different part of the state. If we would have to travel down the history and track actions, it would become difficult to figure out which action changed the visibility. Also, if in the future we were to change the fact that visibility is not dictated by navigation state then not having a separate action for it will break.

Reply:
Why would a change in state ever not change the visibility, or how would we change the state unintentionally? It seems redundant to me that we need to call a visibility change action, it's more of a source of bugs if we _forget_ to call it after the state change. Following the example, we'd have to have an empty action for each operation (maneuver update, distance remaining update, etc).

I still think that we can remove `UpdateVisibility` action and only reduce the available data into a result which is applied, without empty triggers like this.

> Also, if in the future we were to change the fact that visibility is not dictated by navigation state then not having a separate action for it will break.

That would be a breaking change for the entire `component` anyway so that might also be a better point in time to add the necessary action instead of introducing it now without a clear use-case.

Reply:
> it's more of a source of bugs if we forget to call it after the state change. Following the example, we'd have to have an empty action for each operation (maneuver update, distance remaining update, etc).

Agreed with this. We could probably remove the empty `UpdateVisibility`, but we would later have to add a new action UpdateVisibility(show = true/false) if an end user wants to show a view which is hidden or vice versa
cc @LukasPaczos

Reply:
> we would later have to add a new action UpdateVisibility(show = true/false) if an end user wants to show a view which is hidden or vice versa

I still hope this will be controlled by the state variable which would have an understanding of which component should be visible rather than components hardcoding if they are visible for given state (as things currently are).

Reply:
That would be an ideal scenario. In reality the visibility of a view is controlled by many different factors here and is not the same for every view. `Recenter` uses `navigationState` and `cameraState` whereas `Maneuver` uses `navigationState` and `maneuverValue`. It is probably too soon to say if there are any other states/data point that dictate the visibility. Also, it's probably difficult to predict if an end user would just want to show the view irrespective of how it is dictated by the state/data. 

However, we can remove it for now and cross that bridge when we come to it.